controller/backup_controller.go

```diff
@@ -434,99 +510,139 @@ func (bc *BackupController) backupCreation(log logrus.FieldLogger, engineClient
 		biChecksum = bi.Status.Checksum
 	}
 
-	backup.Status.State = types.BackupStateInProgress
-	event(nil, backup.Status.State, backup, volume)
+	var state types.BackupState
+	_, err = engineClient.SnapshotBackup(backup.Name, backup.Spec.SnapshotName, url, biName, biChecksum, backup.Spec.Labels, credential)
+	if err != nil {
+		state = types.BackupStateError
+	} else {
+		state = types.BackupStateInProgress
+	}
+	backup.Status.State = state
+	bc.eventLog(err, state, backup, volume)
+	return err
+}
 
-	go func() {
-		state := backup.Status.State
-		defer func() {
-			backup, err := bc.ds.GetBackup(backup.Name)
-			if err != nil {
-				log.WithError(err).Errorf("Error get backup")
-				return
-			}
-			existingBackup := backup.DeepCopy()
+func (bc *BackupController) hasMonitor(name string) *BackupMonitor {
+	bc.lock.RLock()
+	defer bc.lock.RUnlock()
+	return bc.monitors[name]
+}
 
-			backup.Status.State = state
-			if reflect.DeepEqual(existingBackup.Status, backup.Status) {
-				return
-			}
-			if _, err := bc.ds.UpdateBackupStatus(backup); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
-				log.WithError(err).Errorf("Error updating backup status")
-				return
-			}
-		}()
+func (bc *BackupController) enableBackupMonitor(backup *longhorn.Backup, engineClient engineapi.EngineClient) {
+	monitor := bc.hasMonitor(backup.Name)
+	if monitor != nil {
+		return
+	}
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	bc.monitors[backup.Name] = NewBackupMonitor(bc.logger, backup, engineClient, bc.enqueueBackupForMonitor)
+}
 
-		if _, err = engineClient.SnapshotBackup(backup.Name, backup.Spec.SnapshotName, url, biName, biChecksum, backup.Spec.Labels, credential); err != nil {
-			state = types.BackupStateError
-			event(err, state, backup, volume)
-			return
-		}
+func (bc *BackupController) disableBackupMonitor(backup *longhorn.Backup) {
+	monitor := bc.hasMonitor(backup.Name)
+	if monitor == nil {
+		return
+	}
+	bc.lock.Lock()
+	defer bc.lock.Unlock()
+	delete(bc.monitors, backup.Name)
+	monitor.Close()
+}
 
-		// Monitor snapshot backup progress
-		for {
-			engines, err := bc.ds.ListVolumeEngines(volume.Name)
-			if err != nil {
-				state = types.BackupStateUnknown
-				event(err, state, backup, volume)
-				return
-			}
+const (
+	BackupMonitorTickPeriod = time.Second
+	BackupMonitorSyncPeriod = 2 * time.Second
+)
 
-			bks := &types.BackupStatus{}
-			for _, e := range engines {
-				backupStatusList := e.Status.BackupStatus
-				for _, b := range backupStatusList {
-					if b.SnapshotName == backup.Spec.SnapshotName {
-						bks = b
-						break
-					}
-				}
-			}
-			if bks == nil {
-				state = types.BackupStateUnknown
-				event(err, state, backup, volume)
-				return
-			}
-			if bks.Error != "" {
-				state = types.BackupStateError
-				event(errors.New(bks.Error), state, backup, volume)
-				return
-			}
+type BackupMonitor struct {
```

Comment:
NIT: Consider putting the backup monitor related code into it's own file / package.

In general the backup monitor should be able to be tested independently from the backup controller.
For example via an engine sim client, we currently don't have many unit tests in this regard yet.